### PR TITLE
HBASE-22432 HRegionServer rssStub handling is incorrect and inconsistent

### PR DESCRIPTION
**Pull request by sershe-ms**

### CODE REVIEW

File: hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java [hunk 1]
```diff
@@ -3864,21 +3848,36 @@ public void remoteProcedureComplete(long procId, Throwable error) {
...
+      resetRssStubOnServiceException(rss);
+      throw ProtobufUtil.getRemoteException(se);
+    }
+  }
+
+  private RegionServerStatusService.BlockingInterface ensureRssStub() {
+    return ensureRssStub(false, Integer.MAX_VALUE);
+  }
+
+  private RegionServerStatusService.BlockingInterface ensureRssStub(
```
Comment (Apache9) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java]: In general I think checkClusterConn should always be true. The old implementation for reportProcedureDone seems incorrect, as we can not quit if the rs is shutting down. Maybe we should throw an IOException when the clusterConnection is closed in reportProcedureDone.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 156 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -0 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ master Compile Tests _ |
| +1 | mvninstall | 285 | master passed |
| +1 | compile | 54 | master passed |
| +1 | checkstyle | 76 | master passed |
| +1 | shadedjars | 274 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 195 | master passed |
| +1 | javadoc | 31 | master passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 236 | the patch passed |
| +1 | compile | 51 | the patch passed |
| +1 | javac | 51 | the patch passed |
| +1 | checkstyle | 66 | hbase-server: The patch generated 0 new + 70 unchanged - 1 fixed = 70 total (was 71) |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 267 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 763 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.0.0. |
| +1 | findbugs | 186 | the patch passed |
| +1 | javadoc | 31 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 16992 | hbase-server in the patch failed. |
| +1 | asflicense | 35 | The patch does not generate ASF License warnings. |
| | | 19772 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.TestSplitWALManager |
|   | hadoop.hbase.client.TestSnapshotTemporaryDirectoryWithRegionReplicas |
|   | hadoop.hbase.client.TestFromClientSide |
|   | hadoop.hbase.quotas.TestSpaceQuotas |
|   | hadoop.hbase.master.procedure.TestSCPWithReplicas |
|   | hadoop.hbase.client.TestAdmin1 |
|   | hadoop.hbase.client.TestReplicasClient |
|   | hadoop.hbase.regionserver.TestRegionReplicas |
|   | hadoop.hbase.client.TestFromClientSide3 |
|   | hadoop.hbase.client.TestFromClientSideWithCoprocessor |
|   | hadoop.hbase.master.procedure.TestSCPWithReplicasWithoutZKCoordinated |
|   | hadoop.hbase.replication.multiwal.TestReplicationSyncUpToolWithMultipleAsyncWAL |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-241/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/241 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux 447dcf7d228c 4.4.0-137-generic #163-Ubuntu SMP Mon Sep 24 13:14:43 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | master / 7878389669 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-241/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-241/1/testReport/ |
| Max. process+thread count | 5406 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-241/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java [hunk 2]
```diff
@@ -3858,26 +3864,30 @@ void reportProcedureDone(ReportProcedureDoneRequest request) throws IOException
...
     RegionServerStatusService.BlockingInterface rss = null;
-    while (maxRetries-- >= 0 && (!checkClusterConn
-      || (checkClusterConn && clusterConnection != null && !clusterConnection.isClosed()))) {
-      rss = rssStub;
-      if (rss != null) {
-        break;
+    while (rss == null && maxRetries-- >= 0
+        && (clusterConnection != null && !clusterConnection.isClosed())) {
+      synchronized (rssStubLock) {
+        if (resetRssStub) {
```
Comment (Apache9) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java]: I do not fully understand what's the usage of this `resetRssStub` flag... We do not test it in the condition of the while loop?

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 6 | https://github.com/apache/hbase/pull/241 does not apply to master. Rebase required? Wrong Branch? See https://yetus.apache.org/documentation/in-progress/precommit-patchnames for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hbase/pull/241 |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-241/3/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by HorizonNet** [comment]
@sershe-ms Could you please look into the merge conflict?

**Comment by Apache9** [comment]
Close since the PR is stale and can not be merged. Feel free to reopen.

**Pull request closed without merging**